src/services/EdFi.DataManagementService.Api/Core/ApiSchema/DocumentPath.cs

```diff
@@ -0,0 +1,158 @@
+// SPDX-License-Identifier: Apache-2.0
+// Licensed to the Ed-Fi Alliance under one or more agreements.
+// The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text.Json.Nodes;
+using EdFi.DataManagementService.Api.Core.ApiSchema.Model;
+using EdFi.DataManagementService.Api.Core.Model;
+
+namespace EdFi.DataManagementService.Api.Core.ApiSchema;
+
+/// <summary>
+/// Provides information from a single DocumentPath from the documentPathsMapping portion
+/// of a resourceSchema in an ApiSchema.json document.
+///
+/// For example, a partial documentPathsMapping for a document with a reference to a School document looks like:
+///
+/// "documentPathsMapping": {
+///   "School": {
+///     "isDescriptor": false,
+///     "isReference": true,
+///     "projectName": "Ed-Fi",
+///     "referenceJsonPaths": [
+///       {
+///         "identityJsonPath": "$.schoolId",
+///         "referenceJsonPath": "$.schoolReference.schoolId"
+///       }
+///     ],
+///     "resourceName": "School"
+///   }
+/// }
+///
+/// Here, a DocumentPath object would represent the document object value whose key is "School"
+///
+/// </summary>
+public class DocumentPath(JsonNode _documentPathsNode)
+{
+    private readonly Lazy<bool> _isDescriptor =
+        new(() =>
+        {
+            return _documentPathsNode["isDescriptor"]?.GetValue<bool>()
+                ?? throw new InvalidOperationException(
+                    "Expected isDescriptor to be on DocumentPath, not a reference, or invalid ApiSchema"
```

Comment:
Would it make sense to create an extension method or a private function for getting a value from a `JsonNode`? I see a lot of repeated code here.

But, it is relatively trivial repeated code. If this is the only file that has it, then it would be easy enough to change all of the lines if we wanted a different exception, for example.

My conclusion: could do it, but I don't feel as strongly about it as I would have a few years ago, when "don't repeat yourself" was still a sacred commandment to me. After the `left-pad` npm debacle I've definitely softened up a bit on DRY for utility functions 😆. But if the same functionality were spread across several files, then it still might make sense to have a _local_ shared function, though not a shared _library_.

Reply:
```csharp
public static T Get<T>(this JsonNode node, string key)
{
     return node[key]?.GetValue<T>()
                ?? throw new InvalidOperationException(
                    $"ApiSchema is invalid: Expected {key} to be on DocumentPath."
}
```

Looking carefully as I read down the list, I see that the messages have some nuance to them. Maybe you find a root level, but then you don't find something below it. So the message says "couldn't find this XYZ inside the ABC", not just "i couldn't find XYZ". Easy to introduce that level of detail in the extension method above, but then it doesn't really take you very far compared to reading / maintaining the original.

Reply:
Right. The message differences are valuable in diagnosing the actual problem. Also, many of the return types are record types of ours, so `GetValue<T>()` won't work.